### PR TITLE
An implementation of FGMRES

### PR DESCRIPTION
**Pull request by amontoison**

### CODE REVIEW

**Comment by amontoison** [comment]
```julia
N = LinearOperator(Float64, 20, 20, false, false, (y, x) -> y .= rand(20))  # Random preconditioner
x, stats = fgmres(A, b, N=N)
norm(b - A * x)
1.2601303680448969e-10

**Comment by amontoison** [comment]
@dpo Can you just check the docstring of `FGMRES`?

File: src/fgmres.jl [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+    (x, stats) = fgmres(A, b::AbstractVector{FC}; memory::Int=20,
+                        M=I, N=I, atol::T=√eps(T), rtol::T=√eps(T),
+                        reorthogonalization::Bool=false, itmax::Int=0,
+                        restart::Bool=false, verbose::Int=0, history::Bool=false,
+                        ldiv::Bool=false, callback=solver->false)
+
+`T` is an `AbstractFloat` such as `Float32`, `Float64` or `BigFloat`.
+`FC` is `T` or `Complex{T}`.
+
+Solve the linear system Ax = b using FGMRES method.
```
Comment (dpo) [inline src/fgmres.jl:23]:
```suggestion
Solve the linear system Ax = b using FGMRES.
```

File: src/fgmres.jl [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+
+FGMRES computes a sequence of approximate solutions with the minimal residual property.
```
Comment (dpo) [inline src/fgmres.jl:25]:
```suggestion
FGMRES computes a sequence of approximate solutions with minimum residual.
```

File: src/fgmres.jl [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+FGMRES is a variant of GMRES that allows changes in the right preconditioning at every step.
```
Comment (dpo) [inline src/fgmres.jl:26]:
```suggestion
FGMRES is a variant of GMRES that allows changes in the right preconditioner at each iteration.
```

File: src/fgmres.jl [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+
+This implementation allows a left preconditioner M and a flexible right preconditioner N.
+A situation in which the preconditioner is "not constant" is when a relaxation-type method,
+a Chebyshev iteration or another Krylov subspace method is used as a preconditioner. 
+Compared to GMRES, there is no additional cost incurred in the arithmetic but the memory requirement almost doubles.
+Thus, GMRES is recommended if the right preconditioner N is identical as each iteration.
```
Comment (dpo) [inline src/fgmres.jl:32]:
```suggestion
Thus, GMRES is recommended if the right preconditioner N is constant.
```

File: src/fgmres.jl [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+
+Full reorthogonalization is available with the `reorthogonalization` option.
+
+If `restart = true`, the restarted version FGMRES(k) is used with `k = memory`.
+If `restart = false`, the parameter `memory` should be used as a hint of the number of iterations to limit dynamic memory allocations.
+More storage will be allocated only if the number of iterations exceed `memory`.
```
Comment (dpo) [inline src/fgmres.jl:38]:
```suggestion
More storage will be allocated only if the number of iterations exceeds `memory`.
```

File: src/fgmres.jl [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+
+FGMRES can be warm-started from an initial guess `x0` with the method
```
Comment (dpo) [inline src/fgmres.jl:40]:
```suggestion
FGMRES can be warm-started from an initial guess `x0` with
```

File: src/fgmres.jl [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+  status = "unknown"
+  user_requested_exit = false
+
+  while !(solved || tired || breakdown || user_requested_exit)
+
+    # Initialize workspace.
+    nr = 0  # Number of coefficients stored in Rₖ.
+    for i = 1 : mem
+      V[i] .= zero(FC)  # Orthogonal basis of {Mr₀, MANₖr₀, ..., (MANₖ)ᵏ⁻¹r₀}.
+      Z[i] .= zero(FC)  # Z = [N₁v₁, ..., Nₖvₖ]
```
Comment (amontoison) [inline src/fgmres.jl:172]:
```suggestion
      Z[i] .= zero(FC)  # Zₖ = [N₁v₁, ..., Nₖvₖ]
```

**Pull request merged**